polyfill/lib/calendar.mjs

```diff
@@ -333,53 +332,1354 @@ function resolveNonLunisolarMonth(calendarDate) {
+    if (this.reviseIntlEra) {
+      const { era, eraYear } = this.reviseIntlEra(result, isoDate);
+      result.era = era;
+      result.eraYear = eraYear;
+    }
+    if (this.checkJulianBug) this.checkJulianBug(result, isoDate);
+
+    const calendarDate = this.adjustCalendarDate(result, cache, 'constrain', true);
+    if (calendarDate.year === undefined) throw new RangeError(`Missing year converting ${JSON.stringify(isoDate)}`);
+    if (calendarDate.month === undefined) throw new RangeError(`Missing month converting ${JSON.stringify(isoDate)}`);
+    if (calendarDate.day === undefined) throw new RangeError(`Missing day converting ${JSON.stringify(isoDate)}`);
+    cache.set(key, calendarDate);
+    // Also cache the reverse mapping
+    ['constrain', 'reject'].forEach((overflow) => {
+      const keyReverse = JSON.stringify({
+        func: 'calendarToIsoDate',
+        year: calendarDate.year,
+        month: calendarDate.month,
+        day: calendarDate.day,
+        overflow,
+        id: this.id
+      });
+      cache.set(keyReverse, isoDate);
+    });
+    return calendarDate;
+  },
+  validateCalendarDate(calendarDate) {
+    let { month, year, day, era, eraYear, monthCode, monthExtra } = calendarDate;
+    // When there's a suffix (e.g. "5bis" for a leap month in Chinese calendar)
+    // the derived class must deal with it.
+    if (monthExtra !== undefined) throw new RangeError('Unexpected `monthExtra` value');
+    if (year === undefined && eraYear === undefined) throw new TypeError('year or eraYear is required');
+    if (month === undefined && monthCode === undefined) throw new TypeError('month or monthCode is required');
+    if (day === undefined) throw new RangeError('Missing day');
+    if (eraYear !== undefined && era === undefined && this.hasEra) throw new TypeError('era is required with eraYear');
+    if (era !== undefined && eraYear === undefined && this.hasEra) throw new TypeError('eraYear is required with era');
+    if (monthCode !== undefined) {
+      if (typeof monthCode !== 'string') {
+        throw new RangeError(`monthCode must be a string, not ${ES.Type(monthCode).toLowerCase()}`);
+      }
+      if (!/^M(1?\d)(L?)$/.test(monthCode)) throw new RangeError(`Invalid monthCode: ${monthCode}`);
+    }
+  },
+  /**
+   * Allows derived calendars to add additional fields and/or to make
+   * adjustments e.g. to set the era based on the date or to revise the month
+   * number in lunisolar calendars per
+   * https://github.com/tc39/proposal-temporal/issues/1203.
+   *
+   * The base implementation fills in missing values by assuming the simplest
+   * possible calendar:
+   * - no eras or a constant era defined in `.constantEra`
+   * - non-lunisolar calendar (no leap months)
+   * */
+  adjustCalendarDate(calendarDate /*, cache, overflow, fromLegacyDate = false */) {
+    if (this.calendarType === 'lunisolar') throw new RangeError('Override required for lunisolar calendars');
```

Comment:
Yeah, this used to be used for more but those uses were whittled away over time. That said, I do think it could be useful useful for future readers to know what kind of calendar each of the built-in calendars are, because the behavior of each type can be different, especially for lunisolar calendars.  Knowing the type of calendar can explain a lot about the behavior of each calendar which I found helpful while initially debugging before I memorized each calendar's type.

I also tried to keep the base implementation completely calendar-neutral. AFAIK there's no special-case per-calendar-id code in `nonIsoHelperBase`. I assume this practice is worth continuing. 

But I honestly don't feel that strongly either way. I don't think the current solution is bad, but it'd also be easy to replace that one line of code with some other OOP-y test (e.g. `if (this.needsAdjustCalendarDateOverride)` and remove `calendarType` completely. 

Happy to follow whatever recommendation you have.

Reply:
My instinct was to move it to comments as it's largely dead code, but I think leaving it in is fine too. If you think some version of it is worth keeping, then I trust that and this version's a fine one to keep.

Reply:
The main reason to leave it in would be if this is extended in the future with more lunisolar calendars like Hindu. I'm inclined to leave it in for documentation purposes as well as preventing errors if someone adds Hindu and forgets to override this method (which is required for lunisolar).